spring-cloud-gcp-kms/src/main/java/com/google/cloud/spring/kms/KmsOperations.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2017-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spring.kms;
+
+import com.google.cloud.spring.core.GcpProjectIdProvider;
+
+/**
+ * Describes supported operations that one can perform on the KMS API.
+ *
+ * <p>For the methods you need specify the secret from GCP KMS by URI string.
+ * The following secret URI syntax is supported:
+ *
+ * 1. Shortest form - specify location ID, key ring ID, and key ID. Project is derived through {@link GcpProjectIdProvider}
+ * kms://{location-id}/{key-ring-id}/{key-id}
```

Comment:
Regarding the new `kms://` syntax -- In your personal use-case, did you find the custom URI syntax easier or would you prefer a different kind of interface, maybe something like:

```
KmsTemplate.getKey(projectId, locationId, keyRingId, keyId).encrypt(...)
KmsTemplate.getKey(locationId, keyRingId, keyId).decrypt(...)
```

One of the reasons we chose the URI approach for Secret Manager was because we wanted to let users specify URIs in their `application.properties` files to access secrets directly.

However this use-case doesn't seem to exist in KMS, so I am curious if you would have found dedicated interface methods easier to use (i.e. we don't have to lock into a URI syntax just because we did it in secret manager).

Reply:
Hi @dzou. Indeed the choice of `kms://` was driven by the Secret Manager implementation. Also at the time it seemed nice to have it as uri resource. I am not strong on it and I don't see extra value, my assumption was that it was expected to be like this.

Reply:
The `KmsTemplate` has the functions that do the encryption and decryption and one of the arguments is the encryption Key. Based on the previous comments (`kms` prefix not to be necessary probably redundant) this could be instead the full encryption key as expected from GCP `cryptoKeyName.toString()` (which happens to be called inside the code)

The reason why I choose the template to have a kms key as an argument on its methods vs having a template that has a key preconfigured, had to do with the need of multiple key usage. 
For example applications thats need to encrypt and persist data using different encryption keys, and cases where the key is dynamically specified (KEK/DEK scenarios).

This approach also seems great.

```
KmsTemplate.getKey(projectId, locationId, keyRingId, keyId).encrypt(...)
KmsTemplate.getKey(locationId, keyRingId, keyId).decrypt(...)
```

I am more than happy to proceed on what is best for the project and most suitable for the existing codebase.

Reply:
I see makes sense. I can see the value the URI brings since it is more concise than the `cryptoKeyName.toString()`.

Regarding the URIs, I did some research and found some past projects/work describing people trying to decrypt values in `application.properties` with KMS [[1]](https://github.com/hfgbarrigas/gcp-kms-spring-boot) [[2]](https://stackoverflow.com/questions/57984457/cant-decrypt-property-from-spring-cloud-config-or-environment-variable), so this kind of invalidates what I said earlier; looks like it will be useful to have this URI so that users can specify keys by String in `application.properties` if necessary.

So I think we can move forward with the URIs (sorry for the confusion). I think if we want alternate ways to specify keys we can add it on top in the future too.

Right now our team is trying to reach out to the Google KMS team to get some more input on best practices too. Looking forward to hearing updates from them too.